granulate_utils/metadata/databricks_client.py

```diff
@@ -135,12 +129,50 @@ def _cluster_all_tags_metadata(self) -> Optional[Dict[str, str]]:
         props = env.get("sparkProperties")
         if props is None:
             raise DatabricksJobNameDiscoverException(f"sparkProperties was not found in {env=}")
-        for prop in props:
-            if prop[0] == CLUSTER_TAGS_KEY:
-                try:
-                    all_tags_value = json.loads(prop[1])
-                except Exception as e:
-                    raise DatabricksJobNameDiscoverException(f"Failed to parse {prop=}") from e
-                return {cluster_all_tag["key"]: cluster_all_tag["value"] for cluster_all_tag in all_tags_value}
+        # Creating a dict of the relevant properties and their values.
+        service_name_prop_candidates = {
+            prop[0]: prop[1] for prop in props if (CLUSTER_ALL_TAGS_PROP == prop[0] or CLUSTER_NAME_PROP == prop[0])
+        }
+        if len(service_name_prop_candidates) == 0:
+            # We expect at least one of the properties to be present.
+            raise DatabricksJobNameDiscoverException(f"Failed to create dict of relevant properties {env=}")
+        # First, trying to extract `CLUSTER_TAGS_KEY` property, in case not redacted.
+        if (
+            cluster_all_tags_value := service_name_prop_candidates.get(CLUSTER_ALL_TAGS_PROP)
+        ) is not None and "redacted" not in cluster_all_tags_value:
+            try:
+                cluster_all_tags_value_json = json.loads(cluster_all_tags_value)
+            except Exception as e:
+                raise DatabricksJobNameDiscoverException(f"Failed to parse {cluster_all_tags_value}") from e
+            return self._enforce_pattern(
+                {cluster_all_tag["key"]: cluster_all_tag["value"] for cluster_all_tag in cluster_all_tags_value_json}
+            )
+        # As a fallback, trying to extract `CLUSTER_NAME_PROP` property.
+        elif (cluster_name_value := service_name_prop_candidates.get(CLUSTER_NAME_PROP)) is not None:
+            return self._enforce_pattern({CLUSTER_NAME_KEY: cluster_name_value})
         else:
-            raise DatabricksJobNameDiscoverException(f"Failed to find {CLUSTER_TAGS_KEY=} in {props=}")
+            raise DatabricksJobNameDiscoverException(
+                f"Failed to extract {CLUSTER_ALL_TAGS_PROP} or {CLUSTER_NAME_PROP} from {props=}"
+            )
+
+    @staticmethod
+    def _enforce_pattern(metadata: Dict[str, str]) -> Dict[str, str]:
+        """
+        This function is used to enforce certain regex and other patterns on some metadata values, on which we
+        know problematic to include in service names.
+        """
+        if JOB_NAME_KEY in metadata:
```

Comment:
I don't think it is necessary at this point.
but in case we want to make things more complex:
we can divide the service pattern apply into two families, regex (with re.sub) and `str` manipulations (`.replace()` and `lower()` etc.)
as with string manipulation we'd like to apply on all values I guess.
as with regex, we might want to have a dict of regex compilations to sub, where:
`{<prop key>: <regex>}` 
e.g.: `apply_regex_by_prop_dict = {CLUSTER_NAME_KEY: RUN_ID_REGEX}`